source/_posts/2016-10-24-gsoc-2016-port-nmatrix-to-jruby.markdown

```diff
@@ -0,0 +1,536 @@
+
+## **Code Commits**
+
+[https://github.com/prasunanand/nmatrix/commits/jruby_port](https://github.com/prasunanand/nmatrix/commits/jruby_port)
+
+
+## **Performance**
+We have benchmarked some of the NMatrix functionalities. The following plots compare the performance between NMatrix-JRuby, NMatrix-MRI and NMatrix-MRI using LAPACK/ATLAS libraries.
+
+
+Note:
+
+1. Addition and subtraction are not supported by LAPACK/ATLAS.
+2. NMatrix-MRI relies on LAPACK/ATLAS for calculating determinants and LU Decomposition(lud).
+
+
+![Alt Matrix Addition](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/add.png?raw=true "Fig.3. Matrix Addition")
+![Alt Matrix Subtraction](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/subt.png?raw=true "Fig.4. Matrix Subtraction")
+![Alt Matrix Multiplication](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/mult.png?raw=true "Fig.5. Matrix Multiplication")
+![Alt Gamma operator](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/gamma.png?raw=true "Fig.6. Gamma Operator")
+![Alt Determinant](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/det.png?raw=true "Fig.7. Determinant")
+![Alt LU Facorization](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/plots/lud.png?raw=true "Fig.8. LU Facorization")
+
+**Result:**
+
+1. For, two dimensional matrices, NMatrix-JRuby is currently slower than NMatrix-MRI for matrix multiplication and matrix decomposition functionalities(calculating determinant and factorizing a matrix). NMatrix-JRuby is faster than NMatrix-MRI for other functionalities of a two dimensional matrix like addition, subtraction, trigonometric operations, etc.
+
+2. NMatrix-JRuby is a clear winner when we are working with matrices of arbitrary dimensions.
+
+
+## **Storing n-dimensional matrices as flat arrays**
+
+The major components of NMatrix are shape, elements, dtype and stype. Any nmatrix when initialized, stores the elements as a flat array. ArrayRealVector class is used to store the elements.
+
+`@s` stores elements, `@shape` stores the shape of the NMatrix, while `@dtype` and `@stype` store the data type and storage type respectively. Currently, we have nmatrix-jruby implemented only for double and Ruby object data type.
+
+
+NMatrix-MRI uses `@s` which is an object containing elements, stride and offset (as in C, we need to deal with the memory allocation for the arrays).
+
+![NMatrix](https://github.com/prasunanand/gsoc_blog/blob/master/img/sciruby_blog/nmatrix.png?raw=true "Fig.1. NMatrix")
+
+## **Slicing and Rank**
+
+Implementing slicing was the toughest part of NMatrix-JRuby implementation.
+`NMatrix@s` stores the elements of a matrix as a flat_array. The elements along any dimension are accessed with the help of the stride. `NMatrix#get_stride` calculates the stride with the help of the dimension and shape and returns an array.
```

Comment:
I refer to flat_array when we store the elements in one-dimension array. I used the wrong term.